example/separable_compilation/main.cpp

```diff
@@ -0,0 +1,55 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <Kokkos_Core.hpp>
+
+#include <cstdio>
+#include <iostream>
+
+KOKKOS_RELOCATABLE_FUNCTION void count_even(const long i, long& lcount);
+
+int main() {
+  Kokkos::ScopeGuard scope_guard;
+
+  const long n = 10000;
+
+  std::cout << "Number of even integers from 0 to " << n - 1 << '\n';
+
+  Kokkos::Timer timer;
+  timer.reset();
+
+  // Compute the number of even integers from 0 to n-1, in parallel.
+  long count = 0;
+  Kokkos::parallel_reduce(
+      n, KOKKOS_LAMBDA(const long i, long& lcount) { count_even(i, lcount); },
+      count);
+
+  double count_time = timer.seconds();
+  std::cout << "Parallel: " << count << ", " << count_time << "s\n";
+
+  timer.reset();
+
+  // Compare to a sequential loop.
```

Comment:
Did you try what we discussed that is whether comparing to an inlined version makes any sense at all?

Reply:
RDC:
```
Number of even integers from 0 to 9
Parallel: 5, 0.0176898s
Sequential: 5, 4.769e-06s
Number of even integers from 0 to 99
Parallel: 50, 0.0175152s
Sequential: 50, 5.4222e-05s
Number of even integers from 0 to 999
Parallel: 500, 0.0180551s
Sequential: 500, 0.000495395s
Number of even integers from 0 to 9999
Parallel: 5000, 0.0179687s
Sequential: 5000, 0.00475208s
Number of even integers from 0 to 99999
Parallel: 50000, 0.018883s
Sequential: 50000, 0.0477251s
Number of even integers from 0 to 999999
Parallel: 500000, 0.0211372s
Sequential: 500000, 0.471865s
Number of even integers from 0 to 9999999
Parallel: 5000000, 0.0416765s
Sequential: 5000000, 4.71598s
Number of even integers from 0 to 99999999
Parallel: 50000000, 0.245804s
Sequential: 50000000, 47.2422s
```
no-rdc:
```
Number of even integers from 0 to 9
Parallel: 5, 0.0176316s
Sequential: 5, 4.708e-06s
Number of even integers from 0 to 99
Parallel: 50, 0.017936s
Sequential: 50, 5.9791e-05s
Number of even integers from 0 to 999
Parallel: 500, 0.0175954s
Sequential: 500, 0.000487341s
Number of even integers from 0 to 9999
Parallel: 5000, 0.0177417s
Sequential: 5000, 0.00474439s
Number of even integers from 0 to 99999
Parallel: 50000, 0.0187472s
Sequential: 50000, 0.0472494s
Number of even integers from 0 to 999999
Parallel: 500000, 0.0195275s
Sequential: 500000, 0.472715s
Number of even integers from 0 to 9999999
Parallel: 5000000, 0.0278132s
Sequential: 5000000, 4.72659s
Number of even integers from 0 to 99999999
Parallel: 50000000, 0.110623s
Sequential: 50000000, 47.1561s
```
Up to 10^6 elements there is hardly any difference after that it gets more than twice as slow.

Reply:
meaning there would be a good story to tell

Reply:
For SYCL, there is barely any difference:
> 1.000000e+01 2.832492e-02s (RDC) vs. 1.710768e-02s (inline)
1.000000e+02 1.777367e-02s (RDC) vs. 1.779604e-02s (inline)
1.000000e+03 2.295095e-02s (RDC) vs. 2.303725e-02s (inline)
1.000000e+04 2.068359e-02s (RDC) vs. 2.069026e-02s (inline)
1.000000e+05 2.098447e-02s (RDC) vs. 2.057105e-02s (inline)
1.000000e+06 4.073293e-02s (RDC) vs. 4.044686e-02s (inline)
1.000000e+07 4.990597e-02s (RDC) vs. 4.962549e-02s (inline)
1.000000e+08 1.075163e-01s (RDC) vs. 1.074704e-01s (inline)